### ISSUE TITLE
Investigate reducing the memory usage of the C module

### ISSUE THREAD

**Issue by wilfwilson** [digraphs/Digraphs#433]
...particularly in the Schreier-Sims implementation.

@ChrisJefferson reported in https://github.com/semigroups/Semigroups/issues/634#issuecomment-790661632 that creating a Digraphs stabiliser chain for the first time immediately allocates 1.5 GB of memory, and that moreover, this memory is kept around to be used for any subsequent stabiliser chains. 

(Note that while I did initially do this implementation, it was a long time ago and I have forgotten most of it, so it's somewhat likely that I'm going to misunderstand some stuff.)

Currently, the stabiliser chain, when initialised, mallocs:
* `sizeof(SchreierSims) = 4986904` bytes for the Schreier Sims struct
*  One temporary permutation; permutations in the Schreier Sims data structure require `MAXVERTS * sizeof(uint16_t) = 1024` bytes each
* For each of the `MAXVERTS` levels, a `PermColl` for `MAXVERTS` strong generators, with each `PermColl` requiring `(sizeof(PermColl) +  (MAXVERTS * sizeof(Perm) + (MAXVERTS * new perms))` bytes. So `270540800` bytes in total
* For each of the `MAXVERTS` levels, the space for `2 * MAXVERTS` permutations (for the up to `MAXVERTS` transversal permutations, and their inverses). So `2 * MAXVERTS * MAXVERTS * 1024` bytes, i.e. `536870912` bytes in total.

Add this up and you get about 775 MB. This is half of what Chris said he found, but I've possibly miscounted (@ChrisJefferson?), but in any case, even allocating 775 MB for one stabiliser chain is a **lot** of memory!

**Comment by james-d-mitchell** [comment]
I'm happy for this to be investigated, and support anything sensible that preserves the performance of the homomorphism finding code and possibly reduces the memory requirement. I'm not personally all that worried about allocating a GB of memory, but I can see why some might be worried, and that this might be one cause of the failing CI for the Semigroups package. The key reason we have things arranged like this is for performance, previously (before you wrote the Schreier-Sims in C) we used GAP stabiliser chains and the code spent 98% of its time using those GAP stabiliser chains. The intention was also to avoid repeatedly malloc and freeing memory for the homomorphisms code, so that we can compute many homomorphisms of relatively small graphs without the time penalty for malloc and free. The approach used is probably rather simple minded, I'd be happy to consider another approach!

**Comment by wilfwilson** [comment]
I'm happy for us to continue to prioritise time performance, and not make any tradeoffs in that regard. As you say, and as Chris reassured me, there is no need to worry about the amount that we're allocating, in 64-bit.

Yes, it's a concern in 32-bit mode... but do we care? Perhaps this is a good opportunity to reflect on the 32-bit support of Digraphs. Not to abandon it now, while everything works pretty well, but perhaps to decide how much (if any) effort we want to spend maintaining it; and to decide how and when we might drop this support. (Perhaps we should document that 64-bit is recommended).

I'll think about ways of reducing the Schreier-Sims memory usage without reducing performance. Obviously I'll need to be able to check that any changes don't affect performance: do you feel that the tests (standard/extreme) are extensive enough that their timings can be used to decide whether performance has been changed, one way the another?

**Comment by james-d-mitchell** [comment]
> I'll think about ways of reducing the Schreier-Sims memory usage without reducing performance. Obviously I'll need to be able to check that any changes don't affect performance: do you feel that the tests (standard/extreme) are extensive enough that their timings can be used to decide whether performance has been changed, one way the another?

Yes I think this should be noticeable in the tests, and I'm happy to see what you come up with. I think one major saving could be found by using an upper triangular array (if you see what I mean) rather than a square array to store orbits, transversals, etc in the Schreier-Sims implementation. This would complicate the arithmetic to access entries, but would cut the space required by approx. 50%. Unless I'm misremembering how this works....

**Comment by wilfwilson** [comment]
> I think one major saving could be found by using an upper triangular array (if you see what I mean) rather than a square array to store [..] transversals

Yes this is where I'd start, so the transversals would together be about 256 MB instead of 512 MB.

**Comment by james-d-mitchell** [comment]
This will be resolved when #626 is merged!
